### PR TITLE
Paper reviewing settings

### PR DESCRIPTION
**Pull request by nop33**
Don't forget:
- [x] Update headers in new files by running `python bin/maintenance/update_header.py 2017`

### CODE REVIEW

File: indico/modules/events/papers/controllers/management.py [hunk 1]
```diff
@@ -111,3 +114,28 @@ def _process(self):
...
+    def _process(self):
+        defaults = FormDefaults(content_review_questions=self.event_new.cfp.content_review_questions,
+                                layout_review_questions=self.event_new.cfp.layout_review_questions,
+                                **paper_reviewing_settings.get_all(self.event_new))
+        form = PaperReviewingSettingsForm(event=self.event_new, obj=defaults)
+        if form.validate_on_submit():
+            data = form.data
+            content_review_questions = data.pop('content_review_questions', [])
+            layout_review_questions = data.pop('layout_review_questions', [])
+            self.event_new.paper_review_questions = []
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: That's not how this works ;)

By doing this, SQLAlchemy may delete the existing questions, and thus the associated answers (or try to delete them and fail if no cascading is set up). Either way, it's not what you want.

I see you already added a field similar to the one I added in the abstract dialog, but you need to use logic similar to what I did in `RHManageAbstractReviewing` too (do the assignment to deal with new questions and also - very important - don't commit in case the form wasn't submitted successfully to avoid committing editions/deletions if form validation failed somewhere else).

File: indico/modules/events/papers/fields.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+                                                        **kwargs)
+
+    def process_formdata(self, valuelist):
+        super(PaperReviewQuestionsField, self).process_formdata(valuelist)
+        if valuelist:
+            existing = {x.id: x for x in self.object_data or ()}
+            data = []
+            for pos, entry in enumerate(self.data, 1):
+                question = (existing.pop(int(entry['id']))
+                            if entry.get('id') is not None else PaperReviewQuestionsField())
```
Comment (ThiefMaster) [inline indico/modules/events/papers/fields.py]: wrong class name. you want the model here, not a wtforms field

File: indico/modules/events/papers/fields.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import unicode_literals
+
+from indico.util.i18n import _
+from indico.web.forms.fields import MultipleItemsField
+
+
+class PaperReviewQuestionsField(MultipleItemsField):
```
Comment (ThiefMaster) [inline indico/modules/events/papers/fields.py]: This is almost identical to `AbstractReviewQuestionsField` - let's try to use a common class that handles both or a base class and then a subclass that adds the `no_score` field since the only differences seem to be the name of the question model and the (non-)existence of `no_score`.

One possible way of unifying it in a single class could be a constructor like this:

    ... = ReviewQuestionsField(_('whatever'), question_model=lambda: PaperReviewQuestion(type=PaperReviewType.xxx))

and for abstracts

    ... = ReviewQuestionsField(_('whatever'), question_model=AbstractReviewQuestion,
                               extra_fields=[{'id': 'no_score', 'caption': _("Exclude from score"),
                                              'type': 'checkbox'}])

File: indico/modules/events/papers/controllers/management.py (hunk 1 above)
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: Right now, saving this form will delete all questions for a review type if it has been disabled.

File: indico/modules/events/papers/forms.py [hunk 4]
```diff
@@ -49,3 +51,21 @@ def __init__(self, *args, **kwargs):
...
             del self.layout_reviewers
+
+
+class PaperReviewingSettingsForm(IndicoForm):
+    content_review_questions = PaperReviewQuestionsField(_("Content review questions"))
+    layout_review_questions = PaperReviewQuestionsField(_("Layout review questions"))
+    announcement = IndicoMarkdownField(_('Announcement'), editor=True)
+
+    @generated_data
+    def announcement_render_mode(self):
```
Comment (ThiefMaster) [inline indico/modules/events/papers/forms.py]: Did we have this announcement in legacy? If not, we don't need to store a render mode since it's always markdown. If we did, add the `handle_legacy_description` call like in `RHManageAbstractSubmission`. That way a description that has a HTML render mode (the ZODB migration script would have to set that) will be converted automatically when edited, combined with a warning asking the user to double-check that it still looks fine)

File: indico/modules/events/papers/views.py [hunk 5]
```diff
@@ -16,22 +16,28 @@
 
 from __future__ import unicode_literals
 
+from indico.modules.events.abstracts.views import _MathjaxMixin
```
Comment (ThiefMaster) [inline indico/modules/events/papers/views.py]: should be moved to a more suitable place than `abstracts`, and probably un-underscored

Comment (nop33) [inline indico/modules/events/papers/views.py]: How about `events/util.py`?

File: indico/modules/events/papers/forms.py (hunk 4 above)
Comment (nop33) [inline indico/modules/events/papers/forms.py]: Ah I see. No, we didn't have it in legacy. Thanks!

File: indico/modules/events/papers/views.py (hunk 5 above)
Comment (nop33) [inline indico/modules/events/papers/views.py]: I was about to move it, but then I saw that we import it in the tracks module as well.

Comment (ThiefMaster) [inline indico/modules/events/papers/views.py]: `indico/util/mathjax.py`? it's not directly event-related

File: indico/modules/events/papers/forms.py (hunk 4 above)
Comment (ThiefMaster) [inline indico/modules/events/papers/forms.py]: Then we can get rid of the render_mode and always use `MarkdownText(..)` in the `announcement` property in the `CallForPapers`

File: indico/modules/events/papers/views.py (hunk 5 above)
Comment (nop33) [inline indico/modules/events/papers/views.py]: hm ok

Comment (nop33) [inline indico/modules/events/papers/views.py]: Would you also move the `util.py` from `events/util.py` inside the `events/util/` folder? And maybe rename it to `util/base.py`?

Comment (ThiefMaster) [inline indico/modules/events/papers/views.py]: which events/util folder? also, not a fan of a base.py that's not for base classes

Comment (nop33) [inline indico/modules/events/papers/views.py]: ah sorry, misread sth above

File: indico/modules/events/papers/fields.py (hunk 3 above)
Comment (nop33) [inline indico/modules/events/papers/fields.py]: Seems like we are moving to what @OmeGak was suggesting in the beginning: a `review` module :boom:

Comment (ThiefMaster) [inline indico/modules/events/papers/fields.py]: just for rather generic stuff ;)

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+(function($) {
+    'use strict';
+
+    $.widget('indico.paperemailsettingswidget', {
+        options: {
+            containerElement: null,
+            multipleRecipientOptions: ['notify_on_added_to_event', 'notify_on_assigned_contrib',
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: I don't think this wrapping style is common in JS. I'd go for what we do in other places if it doesn't fit on oneline:
```
foo: [
    ..., ..., ...,
    ...
]
```

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                                       'notify_on_paper_submission'],
+            singleRecipientOptions: ['notify_judge_on_review', 'notify_author_on_judgment']
+        },
+
+        _setUp: function(data) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: not sure about that function name (besides disliking the `setUp` spelling of `setup` it sounds almost like a method that's part of the jquery ui widget system to me)

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            var self = this;
+            var multipleRecipientOptions = self.options.multipleRecipientOptions;
+            var singleRecipientOptions = self.options.singleRecipientOptions;
+
+            for (var i = 0; i < multipleRecipientOptions.length; i++) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: I'd move the `var i;` out since you use it for multiple loops in this function. Easier to see that it's not missing in the next loop. Same for `var key;`

File: indico/util/mathjax.py [hunk 9]
```diff
@@ -1,5 +1,5 @@
 # This file is part of Indico.
-# Copyright (C) 2002 - 2016 European Organization for Nuclear Research (CERN).
```
Comment (ThiefMaster) [inline indico/util/mathjax.py]: I'd amend the commit where you add the file instead of adding a new commit for it

File: indico/modules/events/papers/fields.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+from indico.modules.events.papers.settings import paper_reviewing_settings as settings
+from indico.web.forms.fields import JSONField
+from indico.web.forms.widgets import JinjaWidget
+
+
+class PaperEmailSettingsField(JSONField):
+    CAN_POPULATE = True
+    widget = JinjaWidget('events/papers/forms/paper_email_settings_widget.html')
+
+    def __init__(self, *args, **kwargs):
```
Comment (ThiefMaster) [inline indico/modules/events/papers/fields.py]: not used/needed

File: indico/modules/events/papers/fields.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+#
+# Indico is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+# from __future__ import unicode_literals
```
Comment (ThiefMaster) [inline indico/modules/events/papers/fields.py]: that shouldn't be commented out

File: indico/modules/events/papers/controllers/management.py [hunk 12]
```diff
@@ -226,3 +228,34 @@ def _process(self):
...
+
+class RHManageReviewingSettings(RHManagePapersBase):
+    def _process(self):
+        defaults = FormDefaults(content_review_questions=self.event_new.cfp.content_review_questions,
+                                layout_review_questions=self.event_new.cfp.layout_review_questions,
+                                **paper_reviewing_settings.get_all(self.event_new))
+        form = PaperReviewingSettingsForm(event=self.event_new, obj=defaults)
+        if form.validate_on_submit():
+            data = form.data
+            content_review_questions = data.pop('content_review_questions', [])
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: The whole `*_review_questions` logic here is still not needed. See `RHManageAbstractReviewing`

File: indico/modules/events/papers/controllers/management.py [hunk 13]
```diff
@@ -226,3 +228,34 @@ def _process(self):
...
+                                        event_new=self.event_new)
+                paper_review_questions.append(q)
+            self.event_new.paper_review_questions = paper_review_questions
+
+            for key, value in data.pop('email_settings').iteritems():
+                data[key] = RoleConverter.to_python(value) if isinstance(value, list) else value
+
+            paper_reviewing_settings.set_multi(self.event_new, data)
+            flash(_("The reviewing settings were saved successfully"), 'success')
+            return jsonify_data()
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: no logging?

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js (hunk 7 above)
Comment (nop33) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: I stole it from the `track_role_widget.js`. I'll rename it to `_initilize` or sth.

Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: don't use `_init` though, that one *is* used by the UI core (almost like _create but also runs when calling the constructor again on an element that already has the widget).

maybe `_initFields` or `_populateFields`?

File: indico/modules/events/papers/controllers/management.py (hunk 13 above)
Comment (nop33) [inline indico/modules/events/papers/controllers/management.py]: Well, we don't have logging in the abstract reviewing settings either.

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            singleRecipientOptions: [
+                'notify_judge_on_review', 'notify_author_on_judgment'
+            ]
+        },
+
+        initCheckboxes: function(data) {
+            var self = this;
+            var multipleRecipientOptions = self.options.multipleRecipientOptions;
+            var singleRecipientOptions = self.options.singleRecipientOptions;
+            var i = 0;
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: no need to init

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+            var key = null;
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: same

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+            for (i = 0; i < multipleRecipientOptions.length; i++) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: What about this? More readable IMO

```
multipleRecipientOptions.forEach(function(key) {
    data[key].forEach(function(whatever) {
        $('#{0}-{1}'.format(key, whatever).prop('checked', true);
    });
});
```

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            containerElement: null,
+            multipleRecipientOptions: [
+                'notify_on_added_to_event', 'notify_on_assigned_contrib', 'notify_on_paper_submission'
+            ],
+            singleRecipientOptions: [
+                'notify_judge_on_review', 'notify_author_on_judgment'
+            ]
+        },
+
+        initCheckboxes: function(data) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: can be underscored if not called from outside

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+            var self = this;
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: not really needed

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+                }
+            }
+            for (i = 0; i < singleRecipientOptions.length; i++) {
+                key = singleRecipientOptions[i];
+                $('#' + key).prop('checked', data[key]);
+            }
+        },
+
+        _create: function() {
+            var self = this;
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: not really needed, you only use it once to call `initCheckboxes` but never inside a callback

File: indico/modules/events/papers/controllers/management.py [hunk 20]
```diff
@@ -226,3 +227,26 @@ def _process(self):
...
+                                **paper_reviewing_settings.get_all(self.event_new))
+        form = PaperReviewingSettingsForm(event=self.event_new, obj=defaults)
+        if form.validate_on_submit():
+            data = form.data
+            content_review_questions = data.pop('content_review_questions', [])
+            layout_review_questions = data.pop('layout_review_questions', [])
+            self.event_new.paper_review_questions = content_review_questions + layout_review_questions
+
+            for key, value in data.pop('email_settings').iteritems():
+                data[key] = RoleConverter.to_python(value) if isinstance(value, list) else value
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: Why is this needed?

File: indico/modules/events/papers/templates/forms/paper_email_settings_widget.html [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+                        <th>{% trans %}Layout reviewer{% endtrans %}</th>
+                    {% endif %}
+                    <th>{% trans %}Author{% endtrans %}</th>
+                </tr>
+            </thead>
+            <tbody>
+                <tr class="multiple-recipients">
+                    <td>{% trans %}the person is added to/removed from the conference{% endtrans %}</td>
+                    <td>
+                        <input type="checkbox" name="notify_on_added_to_event" id="notify_on_added_to_event-judge"
```
Comment (OmeGak) [inline indico/modules/events/papers/templates/forms/paper_email_settings_widget.html]: I don't think we are using the `id` for anything. I would leave it out if that's the case. Also, inconsistent underscore/hyphen. There are some more instances of the same in this file.

File: indico/modules/events/papers/controllers/management.py (hunk 20 above)
Comment (nop33) [inline indico/modules/events/papers/controllers/management.py]: cause the value can be a list which needs to be converted to a set. Instead of reimplementing the logic of the `RoleConverter`, I just used it.

Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: Wouldn't it make more sense to do that in the field class? You also pass a set to the form defaults so it'd be more consistent that the field also returns a set

File: indico/modules/events/papers/controllers/management.py [hunk 22]
```diff
@@ -226,3 +227,26 @@ def _process(self):
...
+
+            paper_reviewing_settings.set_multi(self.event_new, data)
+            flash(_("The reviewing settings were saved successfully"), 'success')
+            logger.info("Paper reviewing settings of %s have been updated by %s", self.event_new, session.user)
```
Comment (ThiefMaster) [inline indico/modules/events/papers/controllers/management.py]: `%r`, and for the sake of conciseness I'd skip the "have been" - it goes to a log file after all

File: indico/htdocs/js/indico/jquery/paper_email_settings_widget.js (hunk 18 above)
Comment (nop33) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: In all our jQuery widgets we follow the same principle, I wouldn't change it here.

Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/paper_email_settings_widget.js]: then `var $container = this.options.containerElement;` should probably be changed to use self ;)

File: indico/modules/events/papers/templates/forms/paper_email_settings_widget.html (hunk 21 above)
Comment (ThiefMaster) [inline indico/modules/events/papers/templates/forms/paper_email_settings_widget.html]: I think `notify_on_added_to_event` is supposed to match the name of the field. Seems like a valid case to not follow the usual "dashed" convention. I believe it's used here: https://github.com/indico/indico/pull/2776/files#diff-fe7a46e1cd6f40ed1d66c84b01005e36R42

Comment (ThiefMaster) [inline indico/modules/events/papers/templates/forms/paper_email_settings_widget.html]: However, something that *is* missing here is the field ID. Even though we'll never have two instances of this wtforms field/widget on the same page, I think it's cleaner to have IDs that are guaranteed to be unique nonetheless

Comment (OmeGak) [inline indico/modules/events/papers/templates/forms/paper_email_settings_widget.html]: That's pretty obscure and I bet a hell to debug in the future.

Comment (nop33) [inline indico/modules/events/papers/templates/forms/paper_email_settings_widget.html]: @ThiefMaster you mean I should combine the `{{ field.id }}` with this ID to make sure it's unique?

Comment (ThiefMaster) [inline indico/modules/events/papers/templates/forms/paper_email_settings_widget.html]: yes
